File: src/main/java/pl/lodz/p/it/ssbd2024/ssbd01/entity/mow/Location.java
package pl.lodz.p.it.ssbd2024.ssbd01.entity.mow;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import pl.lodz.p.it.ssbd2024.ssbd01.entity.validator.annotation.ValidLocation;
import pl.lodz.p.it.ssbd2024.ssbd01.util.ControlledEntity;
import pl.lodz.p.it.ssbd2024.ssbd01.util.messages.ExceptionMessages;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Getter
@Setter
@Entity
@Table(name = "location")
@NoArgsConstructor
@ValidLocation
public class Location extends ControlledEntity {

    @Column(nullable = false, unique = true)
    private String name;

    @Column(nullable = false)
    private String street;

    @Column(nullable = false)
    private String buildingNumber;

    @Column(nullable = false)
    private String postalCode;

    @Column(nullable = false)
    private String city;

    @Column(nullable = false)
    private String country;

    @Column(nullable = false)
    private Boolean isActive = true;

    @OneToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE}, mappedBy = "location")
    private List<Room> rooms = new ArrayList<>();

    public Location(String name, String city,String country, String street, String buildingNumber, String postalCode) {
        this.name = name;
        this.street = street;
        this.buildingNumber = buildingNumber;
        this.postalCode = postalCode;
        this.city = city;
        this.country = country;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Location location)) {
            return false;
        }
        return Objects.equals(name, location.name);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(name);
    }
}

File: ssbd202401web/src/types/Location.ts
import { PaginationResponse } from "./Pagination";

export interface Location {
  id: string;
  name: string;
  city: string;
  country: string;
  street: string;
  buildingNumber: string;
  postalCode: string;
}

export interface PaginationLocationResponse
  extends PaginationResponse<Location> {}

export interface CreateLocation {
  name: string;
  city: string;
  country: string;
  street: string;
  buildingNumber: string;
  postalCode: string;
}

export interface LocationsWithNumberOfElements {
  locations: Location[];
  totalElements: number;
}

export interface UpdateLocationDataType {
    name: string;
    city: string;
    country: string;
    street: string;
    buildingNumber: string;
    postalCode: string;
}
